src/sapling/sapling_operation.cpp

```diff
@@ -64,23 +65,41 @@ TxValues calculateTarget(const std::vector<SendManyRecipient>& recipients, const
 
 OperationResult SaplingOperation::build()
 {
+    bool isFromtAddress = false;
+    bool isFromShielded = false;
+
+    if (coinControl) {
+        // if coin control was selected it overrides any other defined configuration
+        std::vector<OutPointWrapper> coins;
+        coinControl->ListSelected(coins);
+        // first check that every selected input is from the same type, cannot be mixed for clear privacy reasons.
+        // error is thrown below if it happens, not here.
+        for (const auto& coin : coins) {
+            if (coin.outPoint.isTransparent) {
+                isFromtAddress = true;
+            } else {
+                isFromShielded = true;
+            }
+        }
+    } else {
+        // Regular flow
+        isFromtAddress = fromAddress.isFromTAddress();
+        isFromShielded = fromAddress.isFromSapAddress();
 
-    bool isFromtAddress = fromAddress.isFromTAddress();
-    bool isFromShielded = fromAddress.isFromSapAddress();
-
-    if (!isFromtAddress && !isFromShielded) {
-        isFromtAddress = selectFromtaddrs;
-        isFromShielded = selectFromShield;
-
-        // It needs to have a from.
         if (!isFromtAddress && !isFromShielded) {
-            return errorOut("From address parameter missing");
+            isFromtAddress = selectFromtaddrs;
+            isFromShielded = selectFromShield;
         }
+    }
 
-        // Cannot be from both
-        if (isFromtAddress && isFromShielded) {
-            return errorOut("From address type cannot be shielded and transparent");
-        }
+    // It needs to have a from.
+    if (!isFromtAddress && !isFromShielded) {
+        return errorOut("From address parameter missing");
+    }
+
+    // Cannot be from both
+    if (isFromtAddress && isFromShielded) {
+        return errorOut("From address type cannot be shielded and transparent");
     }
```

Comment:
Just brainstorming...
This would be clearly a privacy leak, but outright preventing it might negatively impact the UX: imagine a user that wants to send the entire wallet balance in a single transaction, and, at the moment, doesn't care about the privacy. 
Having enough balance but being unable to send because `"From address type cannot be shielded and transparent"` might be quite confusing for non-tech-savvy users...

Since this cannot happen via RPC (the user can select either "from_transparent"/"from_shielded" or a specific address), the only problem is the GUI.
What about if we, instead of failing, present a pop-up that warns the user about the risks and asks confirmation to proceed? 
A simple "Are you sure?" kinda dialog. Like the one we use when spending a cold stake delegation, for example. It could be added in a subsequent PR, of course, this is fine as is now.

Reply:
Why the issue would be the GUI?. 
Same as the RPC, due the clear privacy leak, the GUI is not presenting any flow to craft a transaction with mixed inputs.
If you check the send screen, you can select the funds pot from where the coins are going to be taken. Then the remaining balance is being shown below (transparent or shielded). There is no flow to select transparent and shielded "coins" as inputs of a transaction.

I don't think that enable it would be good in any situation, it's essentially breaking the sole purpose of shield their coins.

--

This error message is more a warning for  someone in the future changing the code or hand crafting the transaction differently. We could make it more friendly and explain the rationale behind if there is a real need for sure.

Reply:
> I don't think that enable it would be good in any situation, it's essentially breaking the sole purpose of shield their coins.

True, but there are situations where the user is not concerned about the privacy, and just wants to "send all".
It was only a from an UX point of view, but no big deal at all, especially now.